internal/controller/services/gateway/gateway_controller_actions.go

```diff
@@ -0,0 +1,180 @@
+/*
+Copyright 2023.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gateway
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	certmanagerv1 "github.com/cert-manager/cert-manager/pkg/apis/certmanager/v1"
+	cmmeta "github.com/cert-manager/cert-manager/pkg/apis/meta/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	gwapiv1 "sigs.k8s.io/gateway-api/apis/v1"
+
+	serviceApi "github.com/opendatahub-io/opendatahub-operator/v2/api/services/v1alpha1"
+	odhtypes "github.com/opendatahub-io/opendatahub-operator/v2/pkg/controller/types"
+)
+
+func createGatewayInfrastructure(ctx context.Context, rr *odhtypes.ReconciliationRequest) error {
+	l := logf.FromContext(ctx).WithName("createGatewayInfrastructure")
+
+	gatewayInstance, ok := rr.Instance.(*serviceApi.Gateway)
+	if !ok {
+		return errors.New("failed to cast the reconciliation request instance to Gateway")
+	}
+	l.Info("Creating Gateway infrastructure", "gateway", gatewayInstance.Name)
+
+	if err := createGatewayClass(rr); err != nil {
+		return fmt.Errorf("failed to create GatewayClass: %w", err)
+	}
+
+	certSecretName, err := handleCertificates(ctx, rr, gatewayInstance)
+	if err != nil {
+		return fmt.Errorf("failed to handle certificates: %w", err)
+	}
+
+	if err := createGateway(rr, gatewayInstance, certSecretName); err != nil {
+		return fmt.Errorf("failed to create Gateway: %w", err)
+	}
+
+	l.Info("Successfully created Gateway infrastructure",
+		"gateway", "odh-gateway",
+		"namespace", gatewayInstance.Spec.Namespace,
+		"domain", gatewayInstance.Spec.Domain,
+		"certificateType", gatewayInstance.Spec.Certificates.Type)
+
+	return nil
+}
+
+func createGatewayClass(rr *odhtypes.ReconciliationRequest) error {
+	gatewayClass := &gwapiv1.GatewayClass{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "odh-gateway-class",
```

Comment:
`odh-gateway-class` is used in more than one place; it is better to declare it as a constant.